codegen/metrics/metrics.go

```diff
@@ -0,0 +1,80 @@
+package metrics
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+)
+
+var (
+	sink *metricSink
+)
+
+func NewSink() {
+	sink = newMetricSink()
```

Comment:
nit: the name `MetricsSink` is also Envoy's name for a metrics sink service, which is a different concept. Maybe we can find a different name for this use case?